### PR TITLE
Introduce ddwaf_(context|subcontext)_multieval to evaluate multiple batches in sequence

### PR DESCRIPTION
**Pull request by Anilm3**
## Overview

This PR adds two new public entrypoints, `ddwaf_context_multieval` and `ddwaf_subcontext_multieval`, that evaluate multiple input batches in sequence within a single call and return one combined result. Functionally each entrypoint behaves like its `_eval` counterpart, except that `data` is an array of maps where every element is treated as a separate input batch and evaluated in order.

To support this, insertions to the object store now result in the batch being added to a queue, which then has to be explicitly applied before it's ready to be evaluated:

- `insert_batch` / `insert_batches` enqueue input (a single map, or an array of maps) without applying it.
- `next_batch` pops and applies the next queued batch, marking its targets as new.
- `flush_input_queue` drains any batches left unevaluated (e.g. after a timeout) on every exit path, applying them as "existing" targets targets so they carry over to subsequent calls and resets the new-target set for the next evaluation.

The evaluation engine now loops over batches, running pre-processors, filters, rules, and post-processors per batch, accumulating events/actions/attributes into the single result.

A new field has been added to the result object:

- **`evaluated`**: an unsigned integer reporting how many batches were fully evaluated. In the normal case this equals the number of non-empty batches; on timeout/error during batch `i` (0-based, counting non-empty batches) it equals `I`, i.e. the index of the batch where the problem occurred. Empty batches are skipped and don't count.

This field is present on the regular `_eval` results too (the result schema is shared), where it's simply `1` for a non-empty evaluated batch or `0` otherwise.

## Files to review (`src/`)

- **`src/object_store.hpp`** / **`src/object_store.cpp`**: core change: the batch queue, `insert_batch`/`insert_batches`/`insert_target`/`next_batch`/`flush_input_queue`/`enqueue_batch`/`apply_batch`, and the `insert_and_apply` test helper. Worth the closest look for the queue lifetime and new-target semantics.
- **`src/evaluation_engine.cpp`**: the per-batch evaluation loop, `flush_input_queue` on exit, and the `evaluated` counter.
- **`src/evaluation_engine.hpp`**: `insert_batch`/`insert_batches` forwarding and the `next_batch`/`insert_and_apply` test internals.
- **`src/interface.cpp`**: the two new public functions plus the array-vs-map dispatch added to `ddwaf_context_eval` / `ddwaf_subcontext_eval`.
- **`src/context.hpp`**: `context`/`subcontext` forwarding for the new store methods and test internals.
- **`src/serializer.hpp`** / **`src/serializer.cpp`**: the `evaluated` field added to `result_components` and `initialise_result_object`.
- **`src/processor/base.hpp`**: processor output switched from `insert` to `insert_target`.

### CODE REVIEW

**Comment by Anilm3** [comment]
@codex review

File: src/evaluation_engine.cpp [hunk 1]
```diff
@@ -42,57 +42,63 @@ void set_context_event_address(object_store &store)
...
+
+                if (should_eval_rules) {
+                    eval_rules(policy, results, deadline);
+                    if (!results.empty()) {
+                        set_context_event_address(store_);
+                    }
                 }
             }
+
+            eval_postprocessors(deadline);
```
Comment (christophe-papazian) [inline src/evaluation_engine.cpp:126]: If eval_preprocessors times out (caught above) and the ruleset has no postprocessors, eval_postprocessors returns without ever checking the deadline — so batches_evaluated still increments. 

Shouldn't there be an early deadline check here, or at the top of eval_postprocessors, to guard against this?

File: tests/integration/interface/context/multieval/test.cpp [hunk 2]
```diff
@@ -0,0 +1,1033 @@
...
+    ddwaf_object data;
+    ddwaf_object_set_map(&data, 0, alloc);
+
+    ddwaf_object result;
+    ddwaf_object_set_invalid(&result);
+
+    EXPECT_EQ(ddwaf_context_multieval(context, &data, alloc, &result, LONG_TIME),
+        DDWAF_ERR_INVALID_OBJECT);
+
+    ddwaf_object_destroy(&data, alloc);
```
Comment (christophe-papazian) [inline tests/integration/interface/context/multieval/test.cpp]: When alloc is non-null, insert_batches moves the object into input_objects_ before the is_array() check, so the context owns the memory even on the INVALID_OBJECT path. Calling ddwaf_object_destroy here would double-free it when the context is later destroyed, no?

File: tests/integration/interface/context/multieval/test.cpp [hunk 3]
```diff
@@ -0,0 +1,1033 @@
...
+    ddwaf_object data;
+    ddwaf_object_set_map(&data, 0, alloc);
+
+    ddwaf_object result;
+    ddwaf_object_set_invalid(&result);
+
+    EXPECT_EQ(ddwaf_subcontext_multieval(subctx, &data, alloc, &result, LONG_TIME),
+        DDWAF_ERR_INVALID_OBJECT);
+
+    ddwaf_object_destroy(&data, alloc);
```
Comment (christophe-papazian) [inline tests/integration/interface/context/multieval/test.cpp]: same here?

File: tests/unit/evaluation_engine_test.cpp [hunk 4]
```diff
@@ -751,10 +751,10 @@ TEST(TestEvaluationEngine, RuleFilterWithSubcontextConditionMatch)
         auto persistent = object_builder_da::map({{"usr.id", "admin"}});
         auto ephemeral = object_builder_da::map({{"http.client_ip", "192.168.0.1"}});
 
-        EXPECT_TRUE(ctx.insert(std::move(persistent)));
+        EXPECT_TRUE(ctx.insert_batch(std::move(persistent)));
```
Comment (christophe-papazian) [inline tests/unit/evaluation_engine_test.cpp]: After this rename from insert() to insert_batch(), persistent is only queued, not applied to targets_. create_subcontext() snapshots targets_ (not the queue), so the subcontext eval runs without usr.id. Shouldn't this use insert_and_apply() to preserve the original behaviour of this test?

**Review by christophe-papazian — COMMENTED** [review]
The serializer now always writes `evaluated` into the result object, but schema/result.json still has `additionalProperties: false` and doesn't list it. Worth updating the schema, or will this silently break consumers doing strict validation?

File: src/evaluation_engine.cpp (hunk 1 above)
Comment (Anilm3) [reply src/evaluation_engine.cpp:126]: It's a bit of an edge-case that probably doesn't matter much as it'll timeout in a rule before reaching the filters. A few other things around the timeout have to be re-worked so I'll address that edge case later.

File: tests/integration/interface/context/multieval/test.cpp (hunk 2 above)
Comment (Anilm3) [reply tests/integration/interface/context/multieval/test.cpp]: This is true, but claude wrote this test without allocating any memory (and the one below).

File: tests/unit/evaluation_engine_test.cpp (hunk 4 above)
Comment (Anilm3) [reply tests/unit/evaluation_engine_test.cpp]: Nice one.

**Review by estringana — APPROVED** [review]

File: src/evaluation_engine.cpp [hunk 5]
```diff
@@ -42,57 +42,63 @@ void set_context_event_address(object_store &store)
...
+            const bool should_eval_filters = should_eval_rules || check_new_filter_targets();
+
+            if (should_eval_filters) {
+                // Filters need to be evaluated even if rules don't, otherwise it'll
+                // break the current condition cache mechanism which requires knowing
+                // if an address is new to this run.
+                const auto &policy = eval_filters(deadline);
+
+                if (should_eval_rules) {
+                    eval_rules(policy, results, deadline);
```
Comment (cataphract) [inline src/evaluation_engine.cpp]: the rfc says "The WAF will evaluate each map sequentially, with the same semantics as a sequence of calls with a single map. However the first _blocking rule triggered_, error or timeout will end the whole evaluation." (emphasis mine). I don't see it stopping here. To be clear, the evaluation of a module inside an eval_rules call will stop that eval_rules call, but the RFC says "whole evaluation", which I take to mean the evaluation of all the maps.

File: src/object_store.hpp [hunk 6]
```diff
@@ -23,9 +23,48 @@ class object_store {
...
+        insert_batch(input);
+        next_batch();
+        return true;
+    }
+
+    // Apply any remaining queued batches as targets *without* marking them as
+    // new (so they won't be evaluated) and clear the new-target set. Invoked
+    // once evaluation finishes, on any exit path including a timeout, mirroring
+    // the inputs being carried over to a subsequent ddwaf_context_eval call.
+    void flush_input_queue();
```
Comment (cataphract) [inline src/object_store.hpp]: not sure bolting this on `object_store` is the correct choice. I'd create a new class

File: src/evaluation_engine.cpp [hunk 7]
```diff
@@ -42,57 +42,63 @@ void set_context_event_address(object_store &store)
...
+                    if (!results.empty()) {
+                        set_context_event_address(store_);
+                    }
```
Comment (cataphract) [inline src/evaluation_engine.cpp]: results is kept across iterations, so this may be called repeatedly

File: src/evaluation_engine.cpp
Comment (cataphract) [inline src/evaluation_engine.cpp]: because the rule_result is not self-contained and the serializer is, perhaps surprisingly, responsible for non-serialization tasks like resolving attributes, I think this can cause attributes to be resolved against the wrong input, in case it's replaced later in the batch. This would not happen if you ran context_val several times in succession

File: src/object_store.cpp [hunk 8]
```diff
@@ -14,26 +14,92 @@
...
+        return false;
+    }
+
+    return insert_batches(array_view{view});
+}
+
+void object_store::enqueue_batch(map_view input)
+{
+    // Batches with no addresses are considered valid as they are harmless
+    if (!input.empty()) {
```
Comment (cataphract) [inline src/object_store.cpp]: this likely has implications for `evaluated` in a way that's confusing for the caller

File: src/evaluation_engine.cpp (hunk 5 above)
Comment (Anilm3) [reply src/evaluation_engine.cpp]: You're right, I missed that.

File: src/object_store.hpp (hunk 6 above)
Comment (Anilm3) [reply src/object_store.hpp]: I was using an "iterator-like" class before but it wasn't much cleaner, I'll give it another thought.

File: src/evaluation_engine.cpp
Comment (Anilm3) [reply src/evaluation_engine.cpp]: I'll see if I can extract the attribute resolution from the serialisation. Otherwise, it's an edge-case we can probably live with...

It would be so much easier if the result was simply an array of results...

File: src/object_store.cpp (hunk 8 above)
Comment (Anilm3) [reply src/object_store.cpp]: I mean... technically there's nothing to evaluate.

Comment (cataphract) [reply src/object_store.cpp]: yes, but the RFC says "Hence, in case of a timeout or an error, it will be the index of the step where the problem occurred."

Comment (Anilm3) [reply src/object_store.cpp]: Addressed in https://github.com/DataDog/libddwaf/pull/494/commits/9b1dbc72cf03275b4135ed88d0180e17d66aef59

File: src/evaluation_engine.cpp
Comment (Anilm3) [reply src/evaluation_engine.cpp]: Addressed in https://github.com/DataDog/libddwaf/pull/494/commits/9b1dbc72cf03275b4135ed88d0180e17d66aef59

File: src/evaluation_engine.cpp (hunk 7 above)
Comment (Anilm3) [reply src/evaluation_engine.cpp]: Addressed in https://github.com/DataDog/libddwaf/pull/494/commits/9b1dbc72cf03275b4135ed88d0180e17d66aef59

File: src/object_store.hpp (hunk 6 above)
Comment (Anilm3) [reply src/object_store.hpp]: Addressed in https://github.com/DataDog/libddwaf/pull/494/commits/9b1dbc72cf03275b4135ed88d0180e17d66aef59

File: src/evaluation_engine.cpp (hunk 5 above)
Comment (Anilm3) [reply src/evaluation_engine.cpp]: Addressed in https://github.com/DataDog/libddwaf/pull/494/commits/9b1dbc72cf03275b4135ed88d0180e17d66aef59

**Comment by Anilm3** [comment]
@codex review

**Review by cataphract — APPROVED** [review]
